### Student Performance Indicator 

#### Life cycle of Machine Learning Project
    - Understanding the problem statement 
    - Data collection 
    - Data checks to perform
    - Exploratory data analysis
    - Data pre-processng 
    - Model training 
    - Choose best model 


#### 1) Probelem Statement :
    - This project understands how the student's performance (test-scores) is affected by the other variables such as gender, ethnicity, etc..

#### 2) Data Collection:
    - kaggle dataset 
    

#### 3) Importing libraries

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
%matplotlib inline

In [5]:
# Reading the dataset 
df = pd.read_csv("StudentsPerformance.csv")

In [8]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [10]:
df.shape

(1000, 8)

#### 3.1) Data Checks to perform

    - Check missing values
    - Check duplicates
    - Check data type 
    - Check he number of unique values of each column 
    - Check statistic of the data-set
    - Check various categories present in the different categorical columns

### Check missing values 

In [14]:
# Check missing values 
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

 - There are no missing values across all 8 columns, meaning no imputation or dropping of rows is required. The dataset is complete and ready for analysis without any data cleaning needed for null values

- ------------------------------------------------------------------------------

#### Check duplicates 

In [19]:
# Check duplicates 
df.duplicated().sum()

0

- No duplicate rows were found, confirming that each of the 1000 records represents a unique student. This ensures our analysis won't be skewed by repeated entries
  

- ------------------------------------------------------------------------------

#### Check data types

In [24]:
# Check nulll and dtypes 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


- All data types are as expected. categorical columns are stored as object and numerical scores as int64. No type conversion is needed before analysis

- ------------------------------------------------------------------------------

#### Checking the number of unique values in each column 


In [29]:
# Unique values in the column 
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

- The low unique counts for categorical columns confirm they are suitable for groupby analysis and can be used as segmentation variables to compare score performance across groups. for example, does completing the test preparation course improve scores?"

- ------------------------------------------------------------------------------

#### Checking the statistics

In [34]:
# Statistics 
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000


##### Insights:
1. Distribution Shape
Compare mean vs median (50%) for each subject — are they close? What does that tell you about symmetry/skew?
- no extreme outliers are pulling the average, so the mean is a reliable measure of typical performance.
  
2. Data Quality / Outliers
Any suspicious values? (hint: that math minimum of 0) — flag it and question it.
- A score of 0 is unlikely in a real exam setting and should be investigated before any further analysis, as it could skew the mean downward.
  
4. Cross-subject Comparison
Which subject has the highest/lowest mean? Which has the most spread (std)? What might that suggest about difficulty or consistency?
- Math std is 15.16, reading is 14.60, writing is 15.19. Writing and math have more variation than reading, meaning reading performance is slightly more consistent across students. 
  
6. Spread & Concentration
Look at the IQR (25%–75% range) — how wide is it? Are students clustered together or all over the place?
- the IQR for math is 20 points (57–77), meaning even the middle 50% of students show considerable variation in performance. This suggests math ability is not consistent across the student population, unlike a subject where students would cluster tightly around the mean.
  
8. So What? (The Takeaway)
End with a conclusion or a question to investigate next. This is what separates a data analyst from someone just reading numbers.
- Math appears to be the most challenging subject with the lowest mean and a potential data quality issue at the minimum. A deeper look into factors like study time or tutoring access could explain the performance gap.

- ------------------------------------------------------------------------------

### Exploring data

In [49]:
df.columns

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [73]:
categorical_features = [features for features in df.columns 
                        if df[features].dtype == "O"]
print(f"Categorical Features : {categorical_features}\n")

numerical_features = [features for features in df.columns 
                      if df[features].dtype !="O"]
print(f"Numerical Features : {numerical_features}")

Categorical Features : ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']

Numerical Features : ['math score', 'reading score', 'writing score']


In [69]:
for feature in categorical_features:
    print(f"Categories in {feature} variable: {df[feature].unique()}")
    print(f"Number of unique categories in {feature} variable: {len(df[feature].unique())}\n")



Categories in gender variable: ['female' 'male']
Number of unique categories in gender variable: 2

Categories in race/ethnicity variable: ['group B' 'group C' 'group A' 'group D' 'group E']
Number of unique categories in race/ethnicity variable: 5

Categories in parental level of education variable: ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Number of unique categories in parental level of education variable: 6

Categories in lunch variable: ['standard' 'free/reduced']
Number of unique categories in lunch variable: 2

Categories in test preparation course variable: ['none' 'completed']
Number of unique categories in test preparation course variable: 2



### Adding columns for "total score" and "average"

In [86]:
df["total score"] = df["math score"] + df["reading score"] + df["writing score"]
df["average"] = df["total score"]/3
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

In [139]:
# Number of students with full marks 
for feature in numerical_features:
    count = (df[feature] == 100).sum()
    print(f"for {feature.split()[0]}: {count} student scored 100")
        

for math: 7 student scored 100
for reading: 17 student scored 100
for writing: 14 student scored 100


In [141]:
# Number of students with marks less than 20 
for feature in numerical_features:
    count = (df[feature] <= 20).sum()
    print(f"for {feature.split()[0]}: {count} students scores less than or equal 20")


for math: 4 students scores less than 20
for reading: 1 students scores less than 20
for writing: 3 students scores less than 20


 - Alot of students struggle in Math, than reading and writing, students do well in reading but fairly in writing